### ISSUE TITLE
Walk method : change in action signature to action (node, parent_node, index, root_node)

### ISSUE THREAD

**Issue by brucou** [joaonuno/tree-model-js#29]
In some traversal cases, it is good to have a direct access to the `parent_node`. The additional parameters can also be useful. This would get the API closer to the `forEach` method for Array.

**Comment by joaonuno** [comment]
The current API allows you to get the parent and the index using `node.parent` and `node.getIndex()`. A possible improvement would be to have the index as an already computed property instead of computing it when `getIndex()` is called. When the tree is initially created you get the root node so you can use that reference if you need it during traversal. Since I'm giving the node to the traversal function, and not the node's model, I can enrich it with properties and don't need to add more function parameters.

**Comment by brucou** [comment]
All true. However there are some good reasons why the api for arrays (forEach, etc.) give back the original array as a parameter. Generally speaking it is not a question of need, there are always ways to do something. I believed this was intended to be a general purpose library (I could not find any with javascript with the same quality as the other languages). If this is just to suit your needs then of course no worries. I reviewed previously the other javascript tree library you mentioned has some of the functionalities, and then you have others. I guess I'll pollute the namespace and do my own. Great demo website by the way.

**Comment by joaonuno** [comment]
> However there are some good reasons why the api for arrays (forEach, etc.) give back the original array as a parameter

I don't doubt that, although to be honest I never had the need to use it. Did you? In what scenarios?

> I believed this was intended to be a general purpose library

It is intended to be general purpose. I just try to keep the API small, whenever it grows I add bugs and maintenance overhead so I really need strong use-cases for adding new methods.

> I could not find any with javascript with the same quality as the other languages

Can you please point me to other languages tree libraries that you consider of superiror quality. I want to learn more about the differences in functionality.

> I guess I'll pollute the namespace and do my own

That is always a choice. I gave you the link to the other library to see if by mixing you had the functionality you needed. Keep in mind that maintaining a library that pleases all is hard. What are the problems you're trying to solve?

Thanks

**Comment by brucou** [comment]
First of all, sorry for the delay in answering, your message somehow made it too low in my mailbox, I just saw it tonight while cleaning old messages. I don't know how to quote past comments but here are a few answers:
1. It is useful to have the original array as third parameter of the typical `.forEach` as in some cases, you will need extra information about the original array (could be length, or access to the previous element), and it allows you to use a pure function instead of a closure. I myself only used that twice for what I remember. 
2. I got your point about maintenance and bugs. I am in the process of finishing my own lib. and it took me two weeks for those very reasons.
3. Java has a few tree libraries (cf. http://stevewedig.com/2014/10/06/digraphs-dags-and-trees-in-java/#library) and with python I like this one : http://www.stat.washington.edu/~hoytak/code/treedict/overview.html#overview (I finally used a similar metaphor in my library). The library with the most features are actually including also the visualization part. Other libraries are into some specific kind of graphs (not just trees, or exotic trees - binary trees etc.).
4. yes I looked closely at the link you provided. Was not general purpose enough for me though. The most important feature for me right now, apart from the usual tree manipulation functions (graft, prune, filter, reduce, find, etc.) is the diff. function. I am managing a component view tree, a la `react`, and having a tree data structure simplifies a lot my code.
